src/rule_definition_tools.jl

```diff
@@ -300,8 +300,25 @@ macro non_differentiable(sig_expr)
 
     primal_name, orig_args = Iterators.peel(sig_expr.args)
 
+    local primal_name_sig
+    # e.g. f(x, y)
+    if primal_name isa Symbol || Meta.isexpr(primal_name, :(.)) ||
+        Meta.isexpr(primal_name, :curly)
+
+        primal_name_sig = :(::Core.Typeof($primal_name))
+        isfunctor = false
+    # e.g. (::T)(x, y)
+    elseif Meta.isexpr(primal_name, :(::))
+        _primal_name = gensym(:primal) 
```

Comment:
Should we make it explicit what primal is?
```suggestion
        _primal_name = gensym(primal_name.args[end]) 
```

Reply:
`_primal_name` is technically just an instance of that type, so I am not sure whether that might be a bit confusing when looking at the expanded code.

Reply:
I have now changed it to `gensym(Symbol(:instance_, primal_name.args[end]))`